msgspec/_core.c

```diff
@@ -2949,7 +3007,18 @@ static PyObject* NamedTupleInfo_Convert(PyObject*);
 static MS_INLINE StructInfo *
 TypeNode_get_struct_info(TypeNode *type) {
     /* Struct types are always first */
-    return type->details[0].pointer;
+    StructInfo *info = type->details[0].pointer;
+#ifdef Py_GIL_DISABLED
+    if (atomic_load(&info->initialized)) {
+        return info;
+    }
+    Py_BEGIN_ALLOW_THREADS
+    /* wait for the StructInfo to be fully initialized by other thread */
+    while (!atomic_load(&info->initialized)) {
+    }
```

Comment:
---

`TypeNode_get_struct_info` spin-loops, which can burn CPU and still races if the pointer becomes visible before initialization.

Fix: Publish the pointer only after the object is fully initialized, then use a single atomic load on the pointer with acquire semantics. No flag, no wait loop:

```c
// writer
StructInfo *tmp = allocate_and_fill(...);            // fully init
atomic_store_explicit(&type->details[0].pointer, tmp, memory_order_release);

// reader
StructInfo *info = atomic_load_explicit(&type->details[0].pointer, memory_order_acquire);
```

If you cannot reorder publication, then use `_Py_yield()` or equivalent inside the loop at minimum, but prefer pointer publication after full init.

Reply:
> TypeNode_get_struct_info spin-loops, which can burn CPU and still races if the pointer becomes visible before initialization.

I don't understand, the load uses seq-cst memory order so there is no data race possible here.
I have tested and verified it using TSAN as well on 3.14 and no data races were there.

Also, this is really an edge case when two threads tries to serialize an object whose typeNode is not fully constructed yet, the performance doesn't matter here as once the node is constructed the fastpath would be taken.